src/number_theory/modular.lean

```diff
@@ -372,14 +487,82 @@ begin
       convert this,
       simp [T] },
     { contrapose! hg',
-      refine ⟨T' * g, by simp [T', matrix.mul, matrix.dot_product, fin.sum_univ_succ], _⟩,
+      refine ⟨T⁻¹ * g, by simp [coe_T_inv, matrix.mul, matrix.dot_product, fin.sum_univ_succ], _⟩,
       rw mul_action.mul_smul,
       have : |(g • z).re - 1| < |(g • z).re| :=
         by cases abs_cases ((g • z).re - 1); cases abs_cases (g • z).re; linarith,
       convert this,
-      simp [T', sub_eq_add_neg] } }
+      simp [coe_T_inv, sub_eq_add_neg] } }
+end
+
+section unique_representative
+
+variables {z}
+
+/-- An auxiliary result en route to `modular_group.c_eq_zero`. -/
+lemma abs_c_le_one (hz : z ∈ 𝒟ᵒ) (hg : g • z ∈ 𝒟ᵒ) : |↑ₘg 1 0| ≤ 1 :=
+begin
+  let c' : ℤ := ↑ₘg 1 0,
+  let c : ℝ := (c' : ℝ),
+  suffices : 3 * c^2 < 4,
+  { rw [← int.cast_pow, ← int.cast_three, ← int.cast_four, ← int.cast_mul, int.cast_lt] at this,
+    replace this : c'^2 ≤ 1^2, { linarith, },
+    rw ← abs_one,
+    exact abs_le_abs_of_sq_le_sq this, },
+  suffices : c ≠ 0 → 9 * c^4 < 16,
+  { rcases eq_or_ne c 0 with hc | hc,
+    { rw hc, norm_num, },
+    { refine (abs_lt_of_sq_lt_sq' _ (by norm_num)).2,
+      specialize this hc,
+      linarith, }, },
+  intros hc,
+  replace hc : 0 < c^4, { rw pow_bit0_pos_iff; trivial, },
+  have h₁ := mul_lt_mul_of_pos_right (mul_lt_mul'' (three_lt_four_mul_im_sq_of_mem_fdo hg)
+      (three_lt_four_mul_im_sq_of_mem_fdo hz) (by linarith) (by linarith)) hc,
+  have h₂ : (c * z.im) ^ 4 / norm_sq (denom ↑g z) ^ 2 ≤ 1 :=
+    div_le_one_of_le (pow_four_le_pow_two_of_pow_two_le
+      (upper_half_plane.c_mul_im_sq_le_norm_sq_denom z g)) (sq_nonneg _),
+  let nsq := norm_sq (denom g z),
+  calc 9 * c^4 < c^4 * z.im^2 * (g • z).im^2 * 16 : by linarith
+           ... = c^4 * z.im^4 / nsq^2 * 16 : by { rw [im_smul_eq_div_norm_sq, div_pow], ring, }
+           ... ≤ 16 : by { rw ← mul_pow, linarith, },
 end
 
+/-- An auxiliary result en route to `modular_group.eq_smul_self_of_mem_fdo_mem_fdo`. -/
+lemma c_eq_zero (hz : z ∈ 𝒟ᵒ) (hg : g • z ∈ 𝒟ᵒ) : ↑ₘg 1 0 = 0 :=
+begin
+  have hp : ∀ {g' : SL(2, ℤ)} (hg' : g' • z ∈ 𝒟ᵒ), ↑ₘg' 1 0 ≠ 1,
+  { intros,
+    by_contra hc,
+    let a := ↑ₘg' 0 0,
+    let d := ↑ₘg' 1 1,
+    have had : T^(-a) * g' = S * T^d, { rw g_eq_of_c_eq_one hc, group, },
+    let w := T^(-a) • (g' • z),
+    have h₁ : w = S • (T^d • z), { simp only [w, ← mul_smul, had], },
+    replace h₁ : norm_sq w < 1 := h₁.symm ▸ norm_sq_S_smul_lt_one (one_lt_norm_sq_T_zpow_smul hz d),
+    have h₂ : 1 < norm_sq w := one_lt_norm_sq_T_zpow_smul hg' (-a),
+    linarith, },
+  have hn : ↑ₘg 1 0 ≠ -1,
+  { intros hc,
+    replace hc : ↑ₘ(-g) 1 0 = 1, { simp [eq_neg_of_eq_neg hc], },
+    replace hg : (-g) • z ∈ 𝒟ᵒ := (neg_smul g z).symm ▸ hg,
+    exact hp hg hc, },
+  specialize hp hg,
+  rcases (int.abs_le_one_iff.mp $ abs_c_le_one hz hg);
+  tauto,
+end
+
+/-- Second Main Fundamental Domain Lemma: if both `z` and `g • z` are in the open domain `𝒟ᵒ`,
```

Comment:
Is this a name used outside mathlib that someone might be searching for? If so, we should make it bold.
```suggestion
/-- Second Main **Fundamental Domain Lemma**: if both `z` and `g • z` are in the open domain `𝒟ᵒ`,
```
If not, fine to leave as is.

Reply:
I'm no number theorist but I don't think this is a standard name.